### PR TITLE
Spacefinder readability improvements

### PR DESCRIPTION
**Pull request by rich-nguyen**
Adding some types to some of the inner functions. Also added `exclusions`, which can be used to track which rule or system excluded a candidate insertion point. Here is what it looks like near the end of `enforceRules()`:

screen shot 2018-02-14 at 16 01 23

### CODE REVIEW

File: static/src/javascripts/projects/common/modules/spacefinder.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 // @flow
 
-// total_hours_spent_maintaining_this = 65
+// total_hours_spent_maintaining_this = 70
```
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/spacefinder.js]: 😏

**Comment by SiAdcock** [comment]
Awesome!

Might need to merge master before merging this PR as it might conflict with the conversion of the `space-filler` tests (#19140)

**Comment by rich-nguyen** [comment]
what are the chances! thanks @SiAdcock

File: static/src/javascripts/projects/common/modules/spacefinder.js [hunk 2]
```diff
@@ -163,68 +181,103 @@ const onInteractivesLoaded = memoize((rules: SpacefinderRules): Promise<
...
     // enforce selector rules
     if (rules.selectors) {
         Object.keys(rules.selectors).forEach(selector => {
-            candidates = candidates.filter(candidate =>
-                testCandidates(
-                    rules.selectors[selector],
-                    candidate,
-                    data.opponents[selector]
-                )
+            exclusions[selector] = [];
```
Comment (JonNorman) [inline static/src/javascripts/projects/common/modules/spacefinder.js]: This is going to make it _so_ much easier 😍

**Review by JonNorman — APPROVED** [review]
Great stuff 👍

**Pull request merged**